src/main/java/io/supertokens/passwordless/Passwordless.java

```diff
@@ -0,0 +1,158 @@
+            @Nullable String userInputCode) throws RestartFlowException, DuplicateLinkCodeHashException,
+            StorageQueryException, NoSuchAlgorithmException, InvalidKeyException {
+        PasswordlessSQLStorage passwordlessStorage = StorageLayer.getPasswordlessStorage(main);
+
+        boolean gotDeviceId = deviceId != null;
+        boolean gotUserInputCode = userInputCode != null;
+
+        if (userInputCode == null) {
+            userInputCode = generateUserInputCode();
+        }
+
+        byte[] deviceIdBytes = new byte[32];
+        String deviceIdHash = null;
+        if (gotDeviceId) {
+            deviceIdBytes = Base64.decodeBase64(deviceId);
+            deviceIdHash = Base64.encodeBase64URLSafeString(Utils.hashSHA256Bytes(deviceIdBytes));
+            PasswordlessDevice device = passwordlessStorage.getDevice(deviceIdHash);
+            if (device == null) {
+                throw new RestartFlowException();
+            }
+        }
+
+        SecureRandom generator = new SecureRandom();
+        while (true) {
+            if (!gotDeviceId) {
+                generator.nextBytes(deviceIdBytes);
+                deviceId = Base64.encodeBase64String(deviceIdBytes);
+                deviceIdHash = Base64.encodeBase64URLSafeString(Utils.hashSHA256Bytes(deviceIdBytes));
+            }
+            String codeId = Utils.getUUID();
+
+            byte[] linkCodeBytes = Utils.hmacSHA256(deviceIdBytes, userInputCode);
+            byte[] linkCodeHashBytes = Utils.hashSHA256Bytes(linkCodeBytes);
+
+            String linkCode = Base64.encodeBase64URLSafeString(linkCodeBytes);
+            String linkCodeHash = Base64.encodeBase64String(linkCodeHashBytes);
+
+            long createdAt = System.currentTimeMillis();
+
+            PasswordlessCode code = new PasswordlessCode(codeId, deviceIdHash, linkCodeHash, createdAt);
+            try {
+                if (!gotDeviceId) {
+                    passwordlessStorage.createDeviceWithCode(email, phoneNumber, code);
+                } else {
+                    passwordlessStorage.createCode(code);
+                }
+                return new CreateCodeResponse(deviceIdHash, codeId, deviceId, userInputCode, linkCode, createdAt);
+            } catch (DuplicateLinkCodeHashException e) {
+                if (gotDeviceId && gotUserInputCode) {
+                    // We only need to rethrow if the user supplied both the deviceId and the userInputCode,
+                    // because in that case the linkCodeHash will always be the same.
+                    throw e;
+                }
+                // It's retrieable otherwise
+            } catch (UnknownDeviceIdHash e) {
+                throw new RestartFlowException();
+            } catch (DuplicateCodeIdException | DuplicateDeviceIdHashException e) {
+                // These are retryable, so ignored here.
+            }
+        }
```

Comment:
The overall code structure can be made better. For example:
- `DuplicateDeviceIdHashException` can only be thrown when using `createDeviceWithCode`, which can only be used when `!gotDeviceId == true`. However, right now, it may suggest that even `createCode` can throw `DuplicateDeviceIdHashException` error.
- Likewise, `UnknownDeviceIdHash` can only be thrown from `createCode`.

I feel like it would be better to make this code structured more around:
```
if (gotDeviceId) {
   while(...) {...}
} else {
   while(...) {...}
}
```